python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -25,6 +26,118 @@
+        _cpu_data_channel: Optional["Channel"] = None,
+    ):
+        self._writer = writer
+        self._readers = readers
+
+        if _gpu_data_channel is not None or _cpu_data_channel is not None:
+            self._gpu_data_channel = _gpu_data_channel
+            self._cpu_data_channel = _cpu_data_channel
+        else:
+            self._gpu_data_channel: TorchTensorNcclChannel = (
+                gpu_data_typ.create_channel(writer, readers)
+            )
+            self._cpu_data_channel: Optional["Channel"] = None
+            if cpu_data_typ is not None:
+                self._cpu_data_channel = cpu_data_typ.create_channel(writer, readers)
+
+        # Used for serialization.
+        self._worker = ray._private.worker.global_worker
+        self._worker.check_connected()
+
+        ctx = ChannelContext.get_current()
+        self.serialization_ctx = ctx.serialization_context
+        assert self.serialization_ctx is not None
+
+    @classmethod
+    def from_channels(
+        cls,
+        gpu_data_channel: "TorchTensorNcclChannel",
+        cpu_data_channel: Optional["Channel"],
+    ):
+        return cls(
+            writer=None,
+            readers=None,
+            gpu_data_typ=None,
+            cpu_data_typ=None,
+            _gpu_data_channel=gpu_data_channel,
+            _cpu_data_channel=cpu_data_channel,
+        )
+
+    def __reduce__(self):
+        return (
+            NestedTorchTensorNcclChannel.from_channels,
+            (self._gpu_data_channel, self._cpu_data_channel),
+        )
+
+    def ensure_registered_as_writer(self):
+        self._gpu_data_channel.ensure_registered_as_writer()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self):
+        self._gpu_data_channel.ensure_registered_as_reader()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_reader()
+
+    def write(self, value: Any):
+        self.serialization_ctx.reset_tensors([])
+
+        try:
+            serialized_cpu_data = self._worker.get_serialization_context().serialize(
+                value
+            )
+        except TypeError as e:
+            sio = io.StringIO()
+            ray.util.inspect_serializability(value, print_file=sio)
+            msg = (
+                "Could not serialize the put value "
+                f"{repr(value)}:\n"
+                f"{sio.getvalue()}"
+            )
+            raise TypeError(msg) from e
+
+        tensors_to_send = self.serialization_ctx.reset_tensors([])
+
+        self._gpu_data_channel.write(tensors_to_send)
+        self._cpu_data_channel.write(serialized_cpu_data)
+
+    def begin_read(self) -> Any:
+        tensors = self._gpu_data_channel.begin_read()
+
+        if self._gpu_data_channel.has_static_type():
+            # If the channel was declared with a static TorchTensorType, then
```

Comment:
> If the channel was declared with a static TorchTensorType, then the task is allowed to return at most one tensor

Would you mind giving me a code reference for this limitation? I don't see `TorchTensorNcclChannel.begin_read` calls the function `has_static_type`.

Reply:
Are you asking how we enforce that only one tensor is sent? It is [here](https://github.com/ray-project/ray/blob/ff4ea691cc8cd123ab5845c391f7d8bd86ef7256/python/ray/experimental/channel/torch_tensor_nccl_channel.py#L277). The writer will check that only one tensor is sent.

Reply:
Thanks! I saw it

```python
            if self.has_static_type():
                # Make sure that there is exactly one tensor to send, and its
                # metadata should have matched the static type.
                if meta_list != [None]:
                    raise ValueError(
                        "DAGNode annotated with "
                        "TorchTensorType(shape=shape, dtype=dtype))` can return at "
                        "most one tensor with the declared `shape` and `dtype`. "
                        "Use TorchTensorType() if value contains more than one "
                        "tensor or tensor of dynamic size."
                    )
```